mat64/io.go

```diff
@@ -5,14 +5,180 @@
 package mat64
 
 import (
+	"bytes"
 	"encoding/binary"
+	"errors"
+
+	"github.com/gonum/matrix"
 )
 
-var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
+// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+const maxLen = int64(int(^uint(0) >> 1))
 
+var (
 	sizeInt64   = binary.Size(int64(0))
 	sizeFloat64 = binary.Size(float64(0))
+
+	errTooBig    = errors.New("mat64: resulting data slice too big")
+	errBadBuffer = errors.New("mat64: data buffer size mismatch")
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  7  number of rows    (int64)
+//   8 - 15  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	bufLen := int64(m.mat.Rows*m.mat.Cols*sizeFloat64 + 2*sizeInt64)
+	if bufLen > maxLen {
```

Comment:
Then I think it is fine to drop this check and replace it with recovering from a panic. After all, we are marshalling a matrix that we already have which means that it is not extremely big so that allocating the destination slice should fail.
We could avoid calling `make`, let `bytes.Buffer` handle growing the buffer and recover `bytes.ErrTooLarge`.